node/subsystem/src/jaeger.rs

```diff
@@ -0,0 +1,244 @@
+    fn default() -> Self {
+        Self {
+            node_name: "unknown_".to_owned(),
+            destination: "http://127.0.0.1:6831".parse().unwrap(),
+        }
+    }
+}
+
+impl JaegerConfig {
+    /// Use the builder pattern to construct a configuration.
+    pub fn builder() -> JaegerConfigBuilder {
+        JaegerConfigBuilder::default()
+    }
+}
+
+
+/// Jaeger configuration builder.
+#[derive(Default)]
+pub struct JaegerConfigBuilder {
+    inner: JaegerConfig
+}
+
+impl JaegerConfigBuilder {
+    /// Set the name for this node.
+    pub fn named<S>(mut self, name: S) -> Self where S: AsRef<str> {
+        self.inner.node_name = name.as_ref().to_owned();
+        self
+    }
+
+    /// Set the recording destination url.
+    pub fn destination<U>(mut self, url: U) -> Self where U: Into<url::Url> {
+        self.inner.destination = url.into();
+        self
+    }
+
+    /// Construct the configuration.
+    pub fn build(self) -> JaegerConfig {
+        self.inner
+    }
+}
+
+/// A wrapper type for a span.
+///
+/// Handles running with and without jaeger.
+pub enum JaegerSpan {
+    /// Running with jaeger being enabled.
+    Enabled(mick_jaeger::Span),
+    /// Running with jaeger disabled.
+    Disabled,
+}
+
+impl JaegerSpan {
+    /// Derive a child span from `self`.
+    pub fn child(&self, name: impl Into<String>) -> Self {
+        match self {
+            Self::Enabled(inner) => Self::Enabled(inner.child(name)),
+            Self::Disabled => Self::Disabled,
+        }
+    }
+    /// Add an additional tag to the span.
+    pub fn add_string_tag(&mut self, tag: &str, value: &str) {
+        match self {
+            Self::Enabled(ref mut inner) => inner.add_string_tag(tag, value),
+            Self::Disabled => {},
+        }
+    }
+}
+
+impl From<Option<mick_jaeger::Span>> for JaegerSpan {
+    fn from(src: Option<mick_jaeger::Span>) -> Self {
+        if let Some(span) = src {
+            Self::Enabled(span)
+        } else {
+            Self::Disabled
+        }
+    }
+}
+
+impl From<mick_jaeger::Span> for JaegerSpan {
+    fn from(src: mick_jaeger::Span) -> Self {
+        Self::Enabled(src)
+    }
+}
+
+/// Shortcut for [`candidate_hash_span`] with the hash of the `Candidate` block.
+#[inline(always)]
+pub fn candidate_hash_span(candidate_hash: &CandidateHash, span_name: impl Into<String>) -> JaegerSpan {
+	hash_span(&candidate_hash.0, span_name)
+}
+
+/// Shortcut for [`hash_span`] with the hash of the `PoV`.
+#[inline(always)]
+pub fn pov_span(pov: &PoV, span_name: impl Into<String>) -> JaegerSpan {
+	hash_span(&pov.hash(), span_name)
+}
+
+/// Creates a `Span` referring to the given hash. All spans created with [`hash_span`] with the
+/// same hash (even from multiple different nodes) will be visible in the same view on Jaeger.
+#[inline(always)]
+pub fn hash_span(hash: &Hash, span_name: impl Into<String>) -> JaegerSpan {
+    INSTANCE.lock().unwrap().span(hash, span_name).into()
```

Comment:
can this lead to a deadlock when used with futures?

Reply:
So, feel free to correct me on this:

The assumption is we run in a deadlock, since the thread this feature is scheduled on blocks on the locked lock but the same type of future is scheduled out on a different threat keeping the lock.

If you have an interrupt roundrobin based executor that uses priority changes of tasks, this might be possible - it depends on the details.
For others (i.e. poll based) this should not be a problem for short lived locks.  After all, a future that is "woken" will execute in a linear fashion until it hits a `task::Poll::{Ready, ..}` or an await point. If there would be await point between thread aquire and release, this could indeed cause dead locks, since the execution is non-linear. For the short lived lock (which does not live across an await point) under the constraint of a non-irq based and thread priority medlying executor, this is safe.

Reply:
If we'll commit this code in production, we need to avoid the `unwrap` and use `expect`

Reply:
Could switch to `parking_lot::Mutex` alternatively